Repair parsing of CHECK_URL in dietpi-software

lwqcz opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready
- [x] Repair parsing of CHECK_URL

**Reference**: N/A

**Commit list/description**:
- I've tried to use `dietpi-software` utility but it was failing to check that my Internet connection is working. The test resulted with `http://[arch=arm64,armhf]: Invalid IPv6 numeric address.`
- The first line of `/etc/apt/sources.list` is `deb [arch=arm64,armhf] https://deb.debian.org/debian/ stretch main contrib non-free`
- Therefore original code failed to find correct URL to test connection on
- I would like to repair this with my piece of code, parsing that file differently
- Anyway I assume that have this connecting checking function be depending on apt's source list is rather non-wise, I would suggest to get the check URL from some kind of separate list or so.

MichaIng committed:
v6.26

+ Backport #3204

MichaIng committed:
v6.26

+ Backport #3204

Commit:
Repair CHECK_URL parameter parsing

Commit:
Repair CHECK_URL parameter parsing (minor typo fix)

MichaIng commented:
@lwqcz 
Many thanks for this. Indeed if the architecture has been added, $2 takes this instead of the wanted URL 🤔.

Your solution on the other hand would fail for ftp:// sources. Of course we can use a more complex sed/mawk/grep combination to reliably get the first sources.list entry, however actually at this stage only network connectivity is to be checked, while all APT sources are checked by G_AGUP (apt update) error-handled anyway.

Hence actually it makes more sense to check an IP only, via ping, which is guaranteed available at any time, e.g. Cloudflares DNS 1.1.1.1 (the in average fastest public DNS, according to some tests I read): `ping -c 1 1.1.1.1`
If we want to test DNS resolving as well (generally makes sense): `ping -c 1 one.one.one.one`

Since generally DNS resolving can fail when raw network connectivity is not given, it would be most helpful for debugging to test raw IP ping first, name resolving afterwards.

What you think?

lwqcz commented:
@MichaIng U R welcome :-)
I totally agree, actually this is not so trivial as it could looks.

Anyway I assume, that `CHECK_URL` routine should check that IP connectivity works, I mean the IP connectivity which matters, therefore it should check for target including DNS resolution as it's needed for APT tools to work properly.

Actually, APT can be used with sources like FTP using just plain IP address, no DNS at all.


Therefore we need to decide if the most important test is testing for:
* the Internet is available
   * `ping -q -W 2000 -i 0.1 -c 10 one.one.one.one 2>&1 | grep "packet loss" | awk -F ', ' '{ print $3 }' | awk -F '%' '{ print $1 }'`
* testing for working network at all
* testing for specific IP/URL defined by default like `one.one.one.one` but easily changeable by the user via network configuration part of `dietpi-config` for instance.

What U think then? 🤔

MichaIng commented:
@lwqcz 
Just as a quick side not for where the current test came from: DietPi did not have some error handling of regular commands in the past, hence it was quite important to assure that the APT source is really available, to avoid running into later issues where an APT package fails to install but configuration to the contained files is attempted. Now all APT related commands are error handled themselves and the script will not continue until each command ran through with exit code `0`.

There is a network time sync done as well, which is still very reasonable, since HTTPS/TLS connections rely on correct system time to not have future dates in certificates, which wouldn't be accepted then. And a general connectivity check first makes sense to not have users digging into their network time sync settings, where actually their network interfaces/adapters need to be checked instead. Since time sync servers can be hostnames as well, name resolution check is reasonable as well.

So the following procedure allows us code-wise and/or the user to check/debug at the right place:
- Check raw network connectivity via always available IP => In case of failure, network interfaces and adapter state needs to be checked.
- Check name resolution via always available hostname => In case of failure, DNS server entry (`/etc/resolv.conf`) needs to be checked, in case DHCP server settings (which shares a DNS server usually) or static IP settings, where `dietpi-config` allows to enter a custom DNS server IP.
- Force network time resync => In case of failure, one now knows that network connectivity and name resolution itself works, hence its the `systemd-timesyncd` service state and/or entered NTP server that needs to be checked.

Later on errors then are mostly assured to be related to the individual called target host and now the own hosts network config.

`G_CHECK_URL` btw is not too good to check general connectivity since it uses wget, hence tests for HTTP/HTTPS/FTP on default ports only. Of course you can manually add a specific port, but wget will then fail, e.g. with empty data error since there is no webserver answering on regular protocol. We do not want to download a file or read data from a HTTP(S) response, hence wget is not the right tool/overkill, for now we simply want to assure that any plain remote host is reachable, and ping/ICMP is the quickest and most generic test to do that, since ping echo responses are done as an integral part of the Linux kernel and do not rely on any additional server software.

I aim to switch from wget to curl for most kind of tests, which is more flexible in some ways, but is not really better then wget for raw connectivity tests.

> Actually, APT can be used with sources like FTP using just plain IP address, no DNS at all.

Yeah it can, but it doesn't make much sense since the main APT servers are CDNs which resolves to a variable closest IP. And who starts to add raw IPs sources.list(.d) or worse check for and replace related 3rd party repo hostnames 😉. Our whole script cannot work without name resolution, hence it makes sense to test this at an early stage.

> ping -q -W 2000 -i 0.1 -c 10 one.one.one.one 2>&1 | grep "packet loss" | awk -F ', ' '{ print $3 }' | awk -F '%' '{ print $1 }'

Does not make sense to send more than a single ping? I mean if a single packet succeeds, then connectivity is assured to be there, if a single packet is lost, then it is reason enough to check what's going on? I'm not sure about situations where some pings go through and some are lost. When done to 1.1.1.1 clearly the issue must be the own network then and such an unreliable connection doesn't seem to be a well situation to start larger download and APT installs based on 🤔. I would keep it simple and send a single ping with default timeouts and grab its exit code.
_____________
```
2019-11-28 14:45:44 root@micha:/var/log# l_message='Checking network connectivity' G_RUN_CMD ping -c 1 1.1.1.1
[  OK  ] Checking network connectivity
2019-11-28 14:46:18 root@micha:/var/log# l_message='Checking network connectivity' G_RUN_CMD ping -c 1 1.12019-11-28 14:46:18 root@micha:/var/log# l_message='Checking network connectivity' G_RUN_CMD ping -c 1 1.1.1.999
[FAILED] ping -c 1 1.1.1.999
        ┌───────────────────────────────┤ DietPi Error Handler: ├────────────────────────────────┐
        │ ping -c 1 1.1.1.999
        │  - Exit code: 2
        │  - DietPi version: v6.26.2 (MichaIng/beta) | HW_MODEL:2 | HW_ARCH:2 | DISTRO:6
        │  - Image creator: DietPi Core Team
        │  - Pre-image: Raspbian Lite
        │
        │ Log file content:
        │ ping: 1.1.1.999: Name or service not known
        │
        │ Unable to continue,  will now terminate.
        │                                                                                        ▒
        │               Retry         : Re-run the last command that failed                      ▒
        │               DietPi-Config : Edit network, APT/NTP mirror settings etc                ▒
        │                                                                                        ▒
        │                                                                                        ↓
        │
        │                        <Ok>                            <Exit>
2019-11-28 14:46:58 root@micha:/var/log# l_message='Checking DNS resolving' G_RUN_CMD ping -c 1 one.one.one.one
[  OK  ] Checking DNS resolving
2019-11-28 14:48:30 root@micha:/var/log# l_message='Checking DNS resolving' G_RUN_CMD ping -c 1 one.one.one.two
[FAILED] ping -c 1 one.one.one.two
        ┌───────────────────────────────┤ DietPi Error Handler: ├────────────────────────────────┐
        │ ping -c 1 one.one.one.two
        │  - Exit code: 2
        │  - DietPi version: v6.26.2 (MichaIng/beta) | HW_MODEL:2 | HW_ARCH:2 | DISTRO:6
        │  - Image creator: DietPi Core Team
        │  - Pre-image: Raspbian Lite
        │
        │ Log file content:
        │ ping: one.one.one.two: Name or service not known
        │
        │ Unable to continue,  will now terminate.
        │                                                                                        ▒
        │               Retry         : Re-run the last command that failed                      ▒
        │               DietPi-Config : Edit network, APT/NTP mirror settings etc                ▒
        │                                                                                        ▒
        │                                                                                        ↓
        │
        │                        <Ok>                            <Exit>
```
- Should do it for now?

lwqcz commented:
@MichaIng the purpose of my multipacket ping was to proof that at least some packets returned within the selected timeout. I seldom use `dietpi` powered on machines on not so stable 3G/LTE/WIFI connection and even when the connection is not very good, APT works rather fine but when the only send packet fails to arrive back (a response I mean) script will be unable to proceed further.
And what about some kind of `ignore-poor-connection` parameter switch? When it'll be used, no connection checks will be made but the outcome can be an error...

MichaIng commented:
@lwqcz 
As you can see above, the error prompt allows to retry, would that be sufficient for your case? I am not too keen to just ignore e.g. 9 of 10 packets failed, as I am not too sure how well all the different setups and download tools can handle such in terms of corrupted downloads etc. Having an error prompt at least gives time to check connection, e.g. move the device closer to the window to enhance mobile signal and things, prior to running e.g. large installs session which can include lots of different downloads via APT, curl, wget from many different sources etc.

Actually I personally would rather do 10 pings and assure that each of them arrived to have a better prove for stable connection 😉. But to keep it simple and sufficient for >95% of cases, a single ping does the job, IMO.

lwqcz commented:
@MichaIng I didn't notice the possibility of retry, then it's all :ok:

MichaIng committed:
v6.27

+ DietPi-Software | Split pure network connectivity test and DNS resolving, to allow better debugging. Do checks against CloudFlare DNS, which is in average the fastest DNS with most servers world-wide and hence guaranteed to be available, including its domain name. The APT mirrors are tested via G_AGUP in an error handled way, hence if this fails, one knows that it is indeed the mirrors and neither DNS resolving, nor connectivity itself. wget and curl both test HTTP(S)/FTP protocol, which is a certain overhead and for debugging reasons not yet wanted here. Those protocols are tested by G_AGUP+G_AGUG and the individual error-handled downloads.

MichaIng commented:
@lwqcz 
Okay I implemented the ping-based test as suggested. Timeout can be controlled via dietpi.txt entry, however a ping should be much faster than any http(s) check. DNS timeout is out of ping command control, I found 20 seconds, Debian manpage states 5 seconds default, however we'll leave it as is for now since this default is valid for all other system calls as well, hence must be sufficient anyway.

Are you okay with this? Then I'd merge.

lwqcz commented:
@MichaIng it seems to be :ok: thanks. But anyway - I do not like hardcoded IP address, what about to have this as an entry from dietpi.txt?

MichaIng commented:
@lwqcz 
Okay, I'll add that later. Worth the effort since we might want to turn those into global functions to use in other scripts (e.g. dietpi-config or dietpi-update) or from console for debugging reasons as well. We can add some targeted error handler steps/solutions/suggestions then, as mentioned further above. However dietpi.txt entries for now to allow IP + domain choice, futher things to come with v6.28.

#### EDIT
Added settings to dev branch: https://github.com/MichaIng/DietPi/commit/9ac8bd9b46feac8d7726e220956f1e7a9fbc7a8e

#### EDIT2
Added commit (below) to use those settings

MichaIng committed:
v6.27

+ DietPi-Software | Connection and DNS resolver testing: Allow to adjust IP/domain to test against via dietpi.txt entries

MichaIng commented:
Okay I'll merge it like this now and run some tests.
Changelog: https://github.com/MichaIng/DietPi/commit/ee1481f6b1001655586fd1dcce7acc935759230f
